lms/views/basic_lti_launch.py

```diff
@@ -278,3 +279,23 @@ def _set_submission_param(self, name, value):
 
         if "submissionParams" in self.context.js_config:
             self.context.js_config["submissionParams"][name] = value
+
+    def _set_focused_user(self, username):
+        """Configure the Hypothesis client to focus on a particular user."""
+
```

Comment:
I'm guessing you'll want this logic to be moved to a service @seanh? We could either move just the logic that calls the API, or go further and move the whole "focus" config generation into a service. Do you have a preference?

Reply:
I think the logic "When it's a SpeedGrader view, then add the "focus" settings to the client config" does belong in the views. So far, although the base client config is generated in `LTILaunchResource`, per-view config customisations have been done in the views.

But I think the logic of how to generate an h userid, and how to generate a client `"focus"` config dict, belongs elsewhere. As does calling the h API.

## Generating the h userid of the focused user

`LTILaunchResource` currently contains a handful of methods for generating properties of Hypothesis users, including the `userid`. But it's hardcoded for generating properties about the currently authenticated user. I think it might be time we extracted a reusable `HUser` out of `LTILaunchResource`:

```python
class HUser:
    username: str
    userid: str
    display_name: str
    provider: str
    provider_unique_id: str
```

`LTILaunchResource` already contains the code necessary to generate the values for all five params from the LTI launch params, in order to create an `HUser` for the authenticated user.

During SpeedGrader launches, to generate an `HUser` for the focused user:

1. `username` is our custom `focused_user` launch param
2. `userid` can be generated from `username` and the `authority` deployment setting
3. `display_name` will have to be retrieved by making a request to the h API. In order to generate a display name from the LTI params like `LTILaunchResource` does we would need the three `lis_person_name_*` params and we don't have those, for the focused user, during a SpeedGrader launch (the launch params will refer to the teacher instead).
4. `provider` is the `tool_consumer_instance_guid` launch param, same as for the authenticated user's `provider`
5. `provider_unique_id` will have to be `None`. Which should be fine as we don't need it. But the inconsistency irks. It can't be generated because it requires the `user_id` LTI param and during a SpeedGrader launch we don't have that for the focused user. And we can't get it from the h API along with `display_name` either, because h's user API doesn't return the `provider_unique_id`.

   We could change the value of the `focused_user` param that we submit to SpeedGrader to be the LTI `user_id` instead of the h `userid`. We'd then be able to generate both the `provider_unique_id` and the `username` from that `focused_user` param (for the `username` we'd also need the `tool_consumer_instance_guid` launch param, which we do have in SpeedGrader launches)

## Calling the h API

I wonder if we need a high-level function or method the current low-level client (the `"hapi"` service) and handle things like generating the URL for the user API, error handling, and default values, keeping these details out of the view.

So far we've been able to keep low-level h API details out of the views by using view decorators: https://github.com/hypothesis/lms/blob/master/lms/views/decorators/h_api.py We have high-level view decorators `upsert_h_user`, `upsert_course_group`, and `add_user_to_group`.

You could add a view decorator that checks for the `focused_user` param, makes the request to the h user API, and adds the display name (or maybe the entire `HUser` object for the focused user) to the `context`.

Or you could add a high-level h API service that wraps the low-level one, or just add a high-level method to the existing low-level service, and have the view call the service.

Reply:
Regarding where the code for calling the h API and inserting the `"focus"` dict into the config would go: I think it would be most in harmony with the current design if this were implemented as a view decorator. Maybe called `focus_user`. Add it to the class's list of view decorators. This decorator would:

1. Do nothing if there's no `"focused_user"` request param

2. Call the h API to get the display name

   Using the existing low-level h API client, and doing error handling, as the existing view decorators do. This is also where defaulting to `"(Couldn't fetch student name)"` would go if you want that.

3. Create an `HUser` for the focused user

4. Generate the `"focus"` dict with the `"username"` and `"displayName"` settings in it, and insert the dict into `context.hypothesis_config`.

Reply:
What are the guidelines for using a service vs a decorator to share functionality between views or extract functionality out of views in the lms app? In h the distinction seems mainly to do with separating concerns that are specific to a particular endpoint (eg. a service call to create an entity in the DB) vs. broader lower-layer concerns applicable to multiple views (but not used outside of views) like logging, or adding CORS headers or configuring how certain categories of response are handled (eg. mapping 404 to 403).

Reply:
Thanks for the feedback. I'll take a look at this tomorrow, going with the proposal to create an `HUser` struct and view decorator. I'm not completely convinced about using decorators for this but it is in keeping with the way that the rest of this view works.

Reply:
No guidelines. The reason those parts were implemented as view decorators was because, when implementing provisioning last year, we didn't want to touch the horrible view code, and view decorators were easy to write separately and then apply to multiple views.

It does turn out to be handy though that you can add the view decorators once in `@view_defaults` but then still turn them off for individual views in `@view_config`.